Fix SIGSEGV while executing lateral subqueries with aggregate functions

whitehawk opened the pull request:
Fix SIGSEGV while executing lateral subqueries with aggregate functions

Problem description:
Some queries with a lateral join cause a segmentation fault. It happens if the
query has a subquery with aggregate functions and without a GROUP BY clause or
it has more than one level nesting.

Cause:
Postgres planner for these queries creates a plan with slices, and the lateral
variable is accessed from a different slice. Lateral variables can't be sent
across motions. So, it causes a segmentation fault.

Fix:
Fix is to force postgres planner to bring such queries to a single query
executor by setting 'force_singleQE' in the PlannerConfig structure. Previously,
this flag was set only for queries with lateral variables and a subquery with
GROUP BY / LIMIT / DISTINCT clauses. Now this condition is complemented by
checking the presence of:
 - aggregate functions;
 - set operations that can cause motion (INTERSECT, EXCEPT, UNION without ALL);
 - recursive CTEs that also can cause motion.
Plus recursive check for all CTEs and SUBQUERY RTEs was added.

whitehawk committed:
Fix segmentation fault with a lateral join

Problem description:
Some queries with lateral join cause segmentation fault. It happens if the query
has nested sub-query (with more than one level nesting) with groupby or limit,
and the lateral variable is accessed from it.

Cause:
Postgres planner for these queries creates a plan with slices, and lateral
variable is accessed from a different slice. Lateral variables can't be send
across motions. So, it causes segmentation fault.

Fix:
Force such queries to single query executor (by setting 'force_singleQE' in
PlannerConfig structure). Previously, this flag was set for queries with lateral
variable and direct sub-query with groupby or limit. This check was not perfect
as the direct sub-query could not contain access to the lateral variable, and it
could have its own sub-query with access to the lateral variable, which was not
checked. Fix is done by improving this check - now
'is_query_contain_limit_groupby' does recursive check for nested sub-queries and
checks that query with groupby or limit really accesses any lateral var.

whitehawk committed:
Refactor is_query_contain_limit_groupby

BenderArenadata commented:
Allure report https://allure.adsw.io/launch/66134

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1136154

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1136155

whitehawk committed:
Refactor 'is_query_contain_limit_groupby' function

BenderArenadata commented:
Failed job Build for ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1140059

whitehawk committed:
Fix segmentation fault for a query with a lateral join and aggregates

BenderArenadata commented:
Failed job Build for ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1140193

BenderArenadata commented:
Allure report https://allure.adsw.io/launch/66240

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1140204

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1140205

BenderArenadata commented:
Failed job Regression tests with ORCA on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1140198

RekGRpth commented on src/backend/optimizer/path/allpaths.c:
```diff
@@ -3003,25 +3003,89 @@ recurse_push_qual(Node *setOp, Query *topquery,
+			return true;
+
+		return false;
+	}
```
How about a little simplification?
```c
	if (IsA(node, Var))
		return equal(node, context);
```

whitehawk replied:
not relevant now

RekGRpth commented on src/backend/optimizer/path/allpaths.c:
```diff
@@ -3003,25 +3003,89 @@ recurse_push_qual(Node *setOp, Query *topquery,
-is_query_contain_limit_groupby(Query *parse)
+is_query_contain_limit_groupby(Query *parse, List *lateral_vars, int levelsup)
 {
+	if (NULL == parse || 0 == list_length(lateral_vars))
```
Can parse be NULL here?

whitehawk replied:
not relevant now

RekGRpth commented on src/backend/optimizer/path/allpaths.c:
```diff
@@ -3003,25 +3003,89 @@ recurse_push_qual(Node *setOp, Query *topquery,
+	{
+		Node	   *node = (Node *) lfirst(lc);
+
+		if (NULL == node || !IsA(node, Var))
```
It is not ORCA-style! Change to
```suggestion
		if (node == NULL || !IsA(node, Var))
```

whitehawk replied:
not relevant now

RekGRpth commented on src/backend/optimizer/path/allpaths.c:
```diff
@@ -3003,25 +3003,89 @@ recurse_push_qual(Node *setOp, Query *topquery,
+		parse->groupClause || parse->distinctClause ||
+		contain_aggs_of_level((Node *)parse->targetList, 0))
+	{
+		if (query_contains_lateral_vars(parse, lateral_vars, levelsup))
```
Maybe we can also combine the conditions as done below?

whitehawk replied:
not relevant now

bimboterminator1 commented on src/backend/optimizer/path/allpaths.c:
```diff
@@ -1700,7 +1700,7 @@ set_subquery_pathlist(PlannerInfo *root, RelOptInfo *rel,
 		 */
 		if ((!bms_is_empty(required_outer)) &&
-			is_query_contain_limit_groupby(subquery))
+			is_query_contain_limit_groupby(subquery, rel->lateral_vars, 1))
```
1.
>'is_query_contain_limit_groupby' checks that the sub-query really accesses at least one lateral var;

Do we really need to check this? Give an example when subquery does not contain any lateral vars when `required_outer` is not empty? Maybe cheking for aggregates is enough?

2.
>'is_query_contain_limit_groupby' checks for presence of aggregate functions
at current level.

If the answer to the first question is afirmative, and we must check lateral vars, then why to check aggregates? How certain can we be that only aggregates are to blame in all the troubles with lateral? May be there are cases besides aggregates that can cause passing params across slices ? Investigate this moment , if possible.  Are there any cases when we have lateral vars without aggregates?

whitehawk replied:
> Maybe cheking for aggregates is enough?

Yes, it looks so. Reworked the patch.

bimboterminator1 replied:
https://github.com/arenadata/gpdb/blob/d136ce0cae8945b413f95df826f2a4b4eb3a0bc4/src/backend/optimizer/path/allpaths.c#L3011

How it takes into account nested subqueries? And under term "aggregates"  in "Maybe cheking for aggregates..." I meant the general concept of queries with group by/limit clause . 
```
explain (costs off) select from (values ('')) _out(b)
join lateral (
    select j from (
        select b as j from t1 
        where t1.b = _out.b group by b
    ) s order by j
) l on true;
                                     QUERY PLAN                                     
------------------------------------------------------------------------------------
 Gather Motion 3:1  (slice2; segments: 3)
   ->  Nested Loop
         ->  Values Scan on "*VALUES*"
         ->  Materialize
               ->  Sort
                     Sort Key: t1.b
                     ->  GroupAggregate
                           Group Key: t1.b
                           ->  Sort
                                 Sort Key: t1.b
                                 ->  Redistribute Motion 3:3  (slice1; segments: 3)
                                       Hash Key: t1.b
                                       ->  GroupAggregate
                                             Group Key: t1.b
                                             ->  Seq Scan on t1
                                                   Filter: (b = "*VALUES*".column1)
 Optimizer: Postgres query optimizer
(17 rows)

```

whitehawk replied:
> How it takes into account nested subqueries?

When I was checking the latest change, I tried to reproduce the issue with nested subqueries, but was not able to do it (thus I removed recursion into RTEs). I guess the reason I couldn't reproduce the issue, while your query can do it, is that I didn't have 'order' clause in select. So, it looks we need to bring the code:
```
	foreach(lc, parse->rtable)
	{
		RangeTblEntry *rte = (RangeTblEntry *) lfirst(lc);

		if (rte->rtekind == RTE_SUBQUERY &&
			is_query_contain_limit_groupby(rte->subquery))
			return true;
	}
```
back.

whitehawk replied:
> And under term "aggregates" in "Maybe cheking for aggregates..." I meant the general concept of queries with group by/limit clause .

Checking for groupby/limit is already there.

whitehawk committed:
Fix SIGSEGV while executing lateral sub-queries with aggregate functions

Problem description:
Some queries with a lateral join cause a segmentation fault. It happens if the
query has a sub-query with aggregate functions and without a groupby clause.

Cause:
Postgres planner for these queries creates a plan with slices, and the lateral
variable is accessed from a different slice. Lateral variables can't be sent
across motions. So, it causes a segmentation fault.

Fix:
Fix is to force postgres planner to bring such queries to a single query
executor by setting 'force_singleQE' in the PlannerConfig structure. Previously,
this flag was set only for queries with lateral variables and a sub-query with
groupby / limit / distinct clauses. Now this condition is complemented by
checking the presence of aggregate functions.

It is rework of previous commit ff57098d0963fa40cf1af2ae7d4fb6d8d229250f

BenderArenadata commented:
Allure report https://allure.adsw.io/launch/66657

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1161589

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1161590

BenderArenadata commented:
Failed job Regression tests with ORCA on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1161583

whitehawk committed:
Bring back recursion into RTEs in is_query_contain_limit_groupby

BenderArenadata commented:
Allure report https://allure.adsw.io/launch/66710

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1164537

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1164538

RekGRpth commented:
I found new issue with patch and gp_cte_sharing = on
```sql
create table t1 (a int, b text, c int) distributed by (a);

insert into t1 values (1, '', 1);

set gp_cte_sharing = on;
set gp_debug_linger = 10;

explain (costs off)
with cte as (select * from t1)
select (select avg(c) from cte where cte.b = tmp.b) from (values ('')) tmp(b) join lateral
(
  select avg(c) from cte where cte.b = tmp.b
) s on true;
with cte as (select * from t1)

select (select avg(c) from cte where cte.b = tmp.b) from (values ('')) tmp(b) join lateral
(
  select avg(c) from cte where cte.b = tmp.b
) s on true;
```
without patch
```sql
                                 QUERY PLAN                                 
----------------------------------------------------------------------------
 Nested Loop
   ->  Values Scan on "*VALUES*"
   ->  Materialize
         ->  Subquery Scan on s
               ->  Aggregate
                     ->  Gather Motion 3:1  (slice2; segments: 3)
                           ->  Aggregate
                                 ->  Subquery Scan on cte
                                       Filter: (cte.b = "*VALUES*".column1)
                                       ->  Shared Scan (share slice:id 2:0)
                                             ->  Materialize
                                                   ->  Seq Scan on t1
   SubPlan 1  (slice0)
     ->  Aggregate
           ->  Subquery Scan on cte_1
                 ->  Result
                       Filter: (t1_1.b = "*VALUES*".column1)
                       ->  Materialize
                             ->  Gather Motion 3:1  (slice1; segments: 3)
                                   ->  Seq Scan on t1 t1_1
 Optimizer: Postgres query optimizer
(21 rows)

with cte as (select * from t1)
select (select avg(c) from cte where cte.b = tmp.b) from (values ('')) tmp(b) join lateral
(
  select avg(c) from cte where cte.b = tmp.b
) s on true;
ERROR:  Error on receive from seg1 slice2 172.19.0.6:6435 pid=40339: server closed the connection unexpectedly
	This probably means the server terminated abnormally
	before or while processing the request.
```
with patch
```sql
FATAL:  Unexpected internal error (cdbmutate.c:2623)
DETAIL:  FailedAssertion("!(plan->flow->flotype == FLOW_SINGLETON)", File: "cdbmutate.c", Line: 2623)
HINT:  Process 43939 will wait for gp_debug_linger=10 seconds before termination.
Note that its locks and other resources will not be released until then.
server closed the connection unexpectedly
	This probably means the server terminated abnormally
	before or while processing the request.
connection to server was lost
```

whitehawk commented:
> I found new issue with patch and gp_cte_sharing = on

If add 'limit', it will fail with the same assert even without the patch:
```
postgres=# explain (costs off)
with cte as (select * from t1)
select (select avg(c) from cte where cte.b = tmp.b) from (values ('')) tmp(b) join lateral
(
  select avg(c) from cte where cte.b = tmp.b limit 1
) s on true;
FATAL:  Unexpected internal error (cdbmutate.c:2623)
DETAIL:  FailedAssertion("!(plan->flow->flotype == FLOW_SINGLETON)", File: "cdbmutate.c", Line: 2623)
HINT:  Process 821845 will wait for gp_debug_linger=10 seconds before termination.
Note that its locks and other resources will not be released until then.
server closed the connection unexpectedly
	This probably means the server terminated abnormally
	before or while processing the request.
The connection to the server was lost. Attempting reset: Succeeded.
postgres=# 
```

bimboterminator1 commented:
The issue with  `gp_cte_sharing = on`  is expected behaviour, this assert does not allow to have shared scans on coordinator.

Commit:
Merge branch 'adb-6.x-dev' into ADBDEV-4858

BenderArenadata commented:
Allure report https://allure.adsw.io/launch/66857

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1173471

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1173473

BenderArenadata commented:
Failed job Regression tests with Postgres on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1173463

RekGRpth commented:
test needs correction

BenderArenadata commented:
Failed job Regression tests with Postgres on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1173464

BenderArenadata commented:
Failed job Regression tests with ORCA on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1173465

BenderArenadata commented:
Failed job Regression tests with ORCA on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1173466

RekGRpth commented:
@whitehawk, test needs correction

whitehawk committed:
Fix test after merge with upstream

BenderArenadata commented:
Allure report https://allure.adsw.io/launch/66926

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1177360

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1177361

RekGRpth commented on src/test/regress/sql/bfv_joins.sql:
```diff
@@ -604,6 +603,10 @@ from
+drop table ttt;
+drop table ttt1;
+
+-- error when use motion deliver a lateral param
```
Why did you remove this line?

whitehawk replied:
'-- error when use motion deliver a lateral param' was removed because queries from the test now doesn't produce any error

RekGRpth commented on src/test/regress/expected/bfv_joins.out:
```diff
@@ -4153,7 +4152,24 @@ from
-ERROR:  Passing parameters across motion is not supported. (cdbmutate.c:3701)
+                               QUERY PLAN                               
+------------------------------------------------------------------------
+ Nested Loop Left Join
```
Why did this case fixed, while next below not?

whitehawk replied:
Fixed the remaining test cases. Also added some more cases.

@RekGRpth , @bimboterminator1 , what do you think about merging test cases from 'goin_gp.sql' to 'dfv_joins.sql'?

RekGRpth replied:
> merging test cases from 'goin_gp.sql' to 'dfv_joins.sql'?

Yes, of course you can move your tests there.

bimboterminator1 replied:
>Also added some more cases.

Is it ok to add new cases in the middle of other patch tests?

whitehawk replied:
> Is it ok to add new cases in the middle of other patch tests?

I've placed new cases closer to similar existing queries. That seems logical. Anyway we are already changing the other patch.

whitehawk committed:
Add additional checks into is_query_contain_limit_groupby

BenderArenadata commented:
Allure report https://allure.adsw.io/launch/67110

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1187266

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1187267

RekGRpth commented on src/backend/optimizer/path/allpaths.c:
```diff
@@ -3006,22 +3006,41 @@ recurse_push_qual(Node *setOp, Query *topquery,
 static bool
 is_query_contain_limit_groupby(Query *parse)
```
May be we should rename this function?

whitehawk replied:
I would agree, now the name is a bit outdated

bimboterminator1 commented on src/backend/optimizer/path/allpaths.c:
```diff
@@ -3006,22 +3006,41 @@ recurse_push_qual(Node *setOp, Query *topquery,
+		parse->hasAggs || parse->hasRecursive)
 		return true;
 
+	// INTERSECT, EXCEPT or UNION without ALL set operations also can
```
Why have you decided that union all can't cause motions?

whitehawk replied:
There are already test cases presented (in join.sql) with lateral and union all, they are planned ok. I couldn't find any example with UNION ALL causing motion that we need to fix. While UNION without ALL will eliminate duplicates, which is a kind of aggregation.

whitehawk committed:
Refactor code and tests for lateral joins with aggregates

BenderArenadata commented:
Allure report https://allure.adsw.io/launch/67112

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1187420

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1187421

RekGRpth commented on src/backend/optimizer/path/allpaths.c:
```diff
@@ -3004,24 +3005,43 @@ recurse_push_qual(Node *setOp, Query *topquery,
+	{
+		CommonTableExpr *cte = (CommonTableExpr *) lfirst(lc);
+
+		if (check_query_for_possible_motion((Query *) cte->ctequery))
+			return true;
```
How about small optimization?
```suggestion
		return check_query_for_possible_motion((Query *) cte->ctequery);
```

whitehawk replied:
But that means if it returned `false`. we will return it to the caller, and skip the remaining checks. Did you try it?

RekGRpth commented on src/backend/optimizer/path/allpaths.c:
```diff
@@ -3004,24 +3005,43 @@ recurse_push_qual(Node *setOp, Query *topquery,
-			 is_query_contain_limit_groupby(lrte->subquery)) ||
-			(rrte->rtekind == RTE_SUBQUERY &&
-			 is_query_contain_limit_groupby(rrte->subquery)))
+		if (rte->rtekind == RTE_SUBQUERY &&
+			check_query_for_possible_motion(rte->subquery))
```
How about small optimization?
```suggestion
		if (rte->rtekind == RTE_SUBQUERY)
			return check_query_for_possible_motion(rte->subquery);
```

RekGRpth commented on src/test/regress/sql/bfv_joins.sql:
```diff
@@ -633,18 +634,93 @@ cross join lateral
   select a+1, b from t1 where a+1 < 10) x;
 
+select * from t1
+cross join lateral
```
May be add some comments about these tests?

whitehawk committed:
Add more comments to bfv_joins test

BenderArenadata commented:
Allure report https://allure.adsw.io/launch/67194

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1192145

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1192146

Commit:
Merge branch 'adb-6.x-dev' into ADBDEV-4858

BenderArenadata commented:
Allure report https://allure.adsw.io/launch/67348

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1200130

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1200131

Commit:
Merge branch 'adb-6.x-dev' into ADBDEV-4858

BenderArenadata commented:
Allure report https://allure.adsw.io/launch/67421

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1204122

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1204123

BenderArenadata commented:
Failed job Regression tests with ORCA on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1204116

RekGRpth commented:
new flaky test?
```diff
DIFF FILE: ../gpdb_src/src/test/regress/regression.diffs
----------------------------------------------------------------------
--- \/home\/gpadmin\/gpdb_src\/src\/test\/regress\/expected\/dispatch\.out	2024-03-25 20:11:13.699629003 +0000
+++ \/home\/gpadmin\/gpdb_src\/src\/test\/regress\/results\/dispatch\.out	2024-03-25 20:11:13.731630533 +0000
@@ -928,7 +928,14 @@
 $_$ language plpgsql;
 -- Query raising 'executor_pre_tuple_processed' error
 select fault_exec_plan(false);
-ERROR:  'executor_pre_tuple_processed' fault triggered
+ fault_exec_plan 
+-----------------
+{
+  'id' => 1,
+  'short' => ''
+}
+GP_IGNORE:(1 row)
+
 -- Same query with explain analyze (should raise
 -- 'executor_pre_tuple_processed' not 'send_exec_stats')
 select fault_exec_plan(true);
@@ -946,11 +953,8 @@
  from gp_segment_configuration where content=-1 and role='p';
  gp_inject_fault 
 -----------------
-{
-  'id' => 1,
-  'short' => 'Success:'
-}
-GP_IGNORE:(1 row)
+ Success:
+(1 row)
 
 select * from gp_segment_configuration a, t13393 ,gp_segment_configuration b where a.dbid = t13393.tc1 limit 0;
  dbid | content | role | preferred_role | mode | status | port | host
name | address | datadir | tc1 | dbid | content | role | preferred_role | mode | status | port | hostname | address | datadir 
```
